parsers/TH.py

```diff
@@ -1,44 +1,148 @@
 MEA_URL = "www.mea.or.th"
 TZ = "Asia/Bangkok"
 
 
-def fetch_EGAT() -> str:
-    # Pull from Min. of Energy for now as I didn't want to write a parser for websocket yet.
-    # The website took around 15s to load, unlike the websocket ver.
-    # Consumption = Production + Import (the latter is done separately)
-    with get(EGAT_GENERATION) as response:
-        EGAT_soup = BeautifulSoup(response.content, "lxml")
+def fetch_production(
+    zone_key: str = "TH",
+    session: Optional[requests.Session] = None,
+    target_datetime: Optional[dt.datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> List[dict]:
+    """Request the last known production mix (in MW) of a given country."""
+    data = _fetch_data(target_datetime, "actual")
+
+    production = []
+    for item in data:
+        production.append(
+            {
+                "zoneKey": zone_key,
+                "datetime": item["datetime"],
+                # All mapped to unknown as there is no available breakdown
+                "production": {"unknown": item["generation"]},
+                "source": EGAT_URL,
+            }
+        )
+    return production
+
 
-    curr_production = EGAT_soup.find_all("p")[6].text.replace(",", "")
-    return curr_production
+def fetch_consumption(
+    zone_key: str = "TH",
+    session: Optional[requests.Session] = None,
+    target_datetime: Optional[dt.datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> List[dict]:
+    """
+    Gets consumption for a specified zone.
+
+    We use the same value as the production for now.
+    But it would be better to include exchanged electricity data if available.
+    """
+    production = fetch_production(target_datetime=target_datetime)
+    consumption = []
+    for item in production:
+        item["consumption"] = item["production"]["unknown"]
+        del item["production"]
+        consumption.append(item)
+    return consumption
+
+
+def fetch_generation_forecast(
+    zone_key: str = "TH",
+    session: Optional[requests.Session] = None,
+    target_datetime: Optional[dt.datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> List[dict]:
+    """Gets generation forecast for specified zone."""
+    data = _fetch_data(target_datetime, "plan")
+
+    production = []
+    for item in data:
+        production.append(
+            {
+                "zoneKey": zone_key,
+                "datetime": item["datetime"],
+                # All mapped to unknown as there is no available breakdown
+                "production": {"unknown": item["generation"]},
+                "source": EGAT_URL,
+            }
+        )
+    return production
+
+
+def _fetch_data(target_datetime: dt.datetime, data_type: str) -> List[dict]:
+    """Fetch actual or planning generation data from the EGAT API endpoint."""
+    url = f"{EGAT_GENERATION_URL}/{data_type}"
+    if target_datetime is None:
+        target_datetime = arrow.now(TZ).datetime
+
```

Comment:
So from my understanding, this would be how that particular section of the code should've looked, right? @mathilde-daugy 
```python
    # Parser runs in UTC; see issue #4407
    if target_datetime is None:
        target_datetime = arrow.utcnow().datetime
```

Reply:
I haven't checked the source or API documentation but assuming the API consume the local datetime and *NOT* UTC I would do something like this and avoid using arrow all together:
```python
    if target_datetime is None:
        target_datetime = datetime.now(tz="<timezone>")
    else:
        target_datetime = target_datetime.astimezone("<timezone>")
```

PS: If we upgrade to Python 3.9+ we can use the native ZoneInfo library for the timezones.

Reply:
Yes, I think their API only supposes the local time. I will try to use this snippet to handle a datetime with a timezone. Thanks!